packages/core/src/components/forms/_file-upload.scss

```diff
@@ -69,21 +70,21 @@ $file-upload-button-width-large: $pt-grid-size * 8.5;
   top: 0;
   right: 0;
   left: 0;
-  margin-right: $file-upload-button-width;
   border-top-right-radius: 0;
   border-bottom-right-radius: 0;
   height: $pt-input-height;
+  padding-right: 0;
   line-height: $pt-input-height;
   color: $pt-text-color-muted;
   user-select: none;
 
-  &::after {
+  // we use ::before rather than ::after although the "Browse" button is visually
+  // positioned to the right since this allows for .pt-file-upload-input to be
+  // paired with .pt-text-overflow-ellipsis if desired.
```

Comment:
a strong reason to use `::after` IIRC (sad there's no comment) is to support icons on the `::before` if the user wishes (icons staked their claim on that pseudoelement on Blueprint day 0).

`.pt-file-upload-input.pt-icon-star` _almost_ works 😮 
![image](https://user-images.githubusercontent.com/464822/29197037-69a7bb2a-7ded-11e7-9d05-2540e109a3be.png)